### PR TITLE
Updated code

### PR DESCRIPTION
**Pull request by theogbob**
contains the updated code for the air drop bot, future updates may include a more "pure random" selection method but works for now.

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -86,12 +132,30 @@ async def cleardb(ctx):
...
+    
+    results = []
+    transaction_explorer_links = []
+    for winner in winners:
+        sendresult = send_coins(rpc_user, rpc_password, rpc_port, rpc_ip, from_address, winner['wallet'], cms_amount)
+        result = sendresult.get('result')
+        results.append(result)
+        transaction_explorer_links.append(f"[Transaction Explorer](https://explorer.cmusic.ai/tx/{result})")
+
+    announcement = f"Congrats! You guy(s) have won the `{cms_amount}` Cmusic airdrop! Your funds should be automatically deposited into the wallets you provided. Here are the winners: "
```
Comment (barrettathome) [inline main.py]:
```suggestion
    announcement = f"Behold the victorious! Below, you'll find the list of winners for the {cms_amount} CmusicAI airdrop. Your funds should be automatically deposited into the wallets you provided. "
```

File: main.py [hunk 2]
```diff
@@ -86,12 +132,30 @@ async def cleardb(ctx):
...
     winners_list = "\n".join([f"{index + 1}) <@{winner['user_id']}> - {winner['wallet']}" for index, winner in enumerate(winners)])
-    announcement = f"Congrats! You have won the Cmusic airdrop! Your funds should be automatically deposited into the wallets you provided. Here are the winners:\n{winners_list} \n\n New airdrop has started, re-enter your wallets!"
+    
+    results = []
+    transaction_explorer_links = []
+    for winner in winners:
+        sendresult = send_coins(rpc_user, rpc_password, rpc_port, rpc_ip, from_address, winner['wallet'], cms_amount)
+        result = sendresult.get('result')
+        results.append(result)
+        transaction_explorer_links.append(f"[Transaction Explorer](https://explorer.cmusic.ai/tx/{result})")
```
Comment (barrettathome) [inline main.py]: Can we make the explorer link an environmental variable.

File: main.py [hunk 3]
```diff
@@ -32,16 +40,63 @@ def has_required_role(ctx):
...
+                else:
+                    await ctx.send("Failed to fetch address information, wallet may not be valid or no transactions [failed explorer check].")
+    else:
+        await ctx.send("Wallet address does not seem valid, [failed regex check].")
+
+@bot.command(name='airdropbal')
+@commands.check(has_required_role)
+async def airdropbal(ctx):
+    balance_response = get_balance(rpc_user, rpc_password, rpc_ip, rpc_port, from_address)
+    await ctx.send("CMS Balance: " + str(balance_response))
```
Comment (barrettathome) [inline main.py]: Can we also set the ticker in the environmental variables.  I'd like to open this up so anyone can use it with any coin.

File: main.py [hunk 4]
```diff
@@ -86,12 +132,30 @@ async def cleardb(ctx):
 
 @bot.command()
 @commands.check(has_required_role)
-async def endairdrop(ctx, num_winners: int):
+async def endairdrop(ctx, num_winners: int, cms_amount: int):
```
Comment (barrettathome) [inline main.py]: Can we update this to use `coin_amount` so it's more generic.

**Review by barrettathome — CHANGES_REQUESTED** [review]
Few updates suggested.

**Comment by theogbob** [comment]
yep, ill add these in a bit

File: main.py (hunk 2 above)
Comment (theogbob) [reply main.py]: done

**Comment by theogbob** [comment]
Done

File: main.py [hunk 5]
```diff
@@ -32,16 +42,63 @@ def has_required_role(ctx):
...
-        async with session.get(f"https://explorer.cmusic.ai/ext/getaddress/{address}") as response:
-            if response.status == 200:
-                data = await response.json()
-                json_data = json.dumps(data, indent=2)
-                await ctx.send("```json\n" + json_data + "```")
-            else:
-                await ctx.send("Failed to fetch address information.")
+    if pattern.match(address):
+        async with aiohttp.ClientSession() as session:
+            async with session.get(f"https://explorer.cmusic.ai/ext/getaddress/{address}") as response:
```
Comment (barrettathome) [inline main.py]: Need to use the explorer env variable here.

File: main.py [hunk 6]
```diff
@@ -52,18 +109,9 @@ async def on_message(message):
             if any(entry["user_id"] == user_id and entry["wallet"] == message.content for entry in db):
                 await message.author.send("Sorry, you have already entered the Cmusic Airdrop!")
```
Comment (barrettathome) [inline main.py]: We should strip out any Cmusic references and put them in an env variable.  Maybe call it something like `project_name`.

File: main.py [hunk 7]
```diff
@@ -52,18 +109,9 @@ async def on_message(message):
...
-                                await message.author.send("Oops, Are you sure that's a Cmusic wallet? Please verify that it is the correct wallet and try again.")
-                            else:
-                                db.append({"user_id": user_id, "wallet": message.content})
-                                save_db(db)
-                                await message.author.send("Congrats! You have successfully entered into the Cmusic Airdrop!")
-                        else:
-                            await message.author.send("Failed to fetch address information.")
+                db.append({"user_id": user_id, "wallet": message.content})
+                save_db(db)
+                await message.author.send("Congrats! You have successfully entered into the Cmusic Airdrop!")
```
Comment (barrettathome) [inline main.py]: Same here

File: main.py [hunk 8]
```diff
@@ -52,18 +109,9 @@ async def on_message(message):
...
         else:
             await message.author.send("Oops, Are you sure that's a Cmusic wallet? Please verify that it is the correct wallet and try again.")
```
Comment (barrettathome) [inline main.py]: And here

File: main.py [hunk 9]
```diff
@@ -86,12 +134,30 @@ async def cleardb(ctx):
...
+    
+    results = []
+    transaction_explorer_links = []
+    for winner in winners:
+        sendresult = send_coins(rpc_user, rpc_password, rpc_port, rpc_ip, from_address, winner['wallet'], coin_amount)
+        result = sendresult.get('result')
+        results.append(result)
+        transaction_explorer_links.append(f"[Transaction Explorer]({eurl}{result})")
+
+    announcement = f"Behold the victorious! Below, you'll find the list of winners for the {cms_amount} `{ticker}` airdrop. Your funds should be automatically deposited into the wallets you provided. "
```
Comment (barrettathome) [inline main.py]: Reference to `cms_amount`

File: main.py [hunk 10]
```diff
@@ -86,12 +134,30 @@ async def cleardb(ctx):
...
+
+    winners_list_with_links = "\n".join([f"{index + 1}) <@{winner['user_id']}> - {winner['wallet']} {link}" for index, (winner, link) in enumerate(zip(winners, transaction_explorer_links))])
+
+    announcement += f"\n\n {winners_list_with_links}\n\n `New airdrop has begun! Re-Enter your wallets to participate` @everyone`!`"
```
Comment (barrettathome) [inline main.py]:
```suggestion
    announcement += f"\n\n {winners_list_with_links}\n\n `New airdrop has begun! Re-Enter your wallet address to participate` @everyone`!`"
```

File: main.py [hunk 11]
```diff
@@ -6,16 +6,26 @@
...
+rpc_port = os.environ.get("rpc_port")
+rpc_ip = os.environ.get("rpc_ip")
+from_address = os.environ.get("from_address")
+eurl = os.environ.get("explorer_url")
+ticker = os.environ.get("ticker")
+
 intents = discord.Intents.all()
 bot = commands.Bot(command_prefix='$>', intents=intents)
 
 pattern = re.compile(r"^(C[a-zA-Z0-9]{33})$")
```
Comment (barrettathome) [inline main.py]: We should update this to maybe an env variable so others can use it with different wallet addresses.  Maybe add in a link for the readme on how to generate REGEX for their particular wallet.

**Review by barrettathome — CHANGES_REQUESTED** [review]
NAB but can we add in a readme file on how to use this?  At the very least we can link to the discord where they can contact you on how to set this up on their own server.

**Comment by theogbob** [comment]
Fixed them all, updated README aswell

**Review by barrettathome — APPROVED** [review]
Looks good, nice work!

**Pull request merged**